### ISSUE TITLE
Do we want typedef support?

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#406]
Because I do lot of API declares, I've sometime wished that VB* would allow me to create typedefs -- having `LongPtr` represent `HWND`, `HDC`, and so forth can get quite tedious and make it possible to accidentally mix wrong kind of data that are otherwise compatible. If we allowed for capability to create a `typedef` statement, that would enable the compiler to provide a warning when a suspicious cast from one `LongPtr` to another `LongPtr` is going to be performed, which would go long way toward eliminating this class of programming errors. We can then specify explicit cast where it is appropriate which requires language support for this new class of type casting.

However, I'm on the fence over this one - the `typedef` statement is very Cish and not very BASICish. We have UDT but single-member UDT is quite awkward to use. This would be not very useful in a codebase that doesn't deal with API as much and thus would potentially be just a feature bloat. 

It's also possible there are better way to solve this problem that doesn't involve adding even more lines of declarations to define the data types before the `Declare`s. 

Thoughts?

**Comment by WaynePhillipsEA** [comment]
I would be in favour of this.  One of my goals with this project is to become the best possible COM / OLE Automation tool, and as such we should bear in mind that COM type libraries offer full support for typedefs ('aliases'), yet we currently have no way to create a type library that exposes an alias.   Type library aliases are already supported in tB for consumption, and so this addition would be relatively simple to implement if we can agree a suitable syntax.

I'd be interested to hear others thoughts on this.  Particularly from those that work with low-level Win API code often, which is where typedefs are often encountered.

**Comment by XusinboyBekchanov** [comment]
FreyBasik uses Type for this:
`Type PControl As Control Ptr`

**Comment by XusinboyBekchanov** [comment]
FreeBasic uses Type for this:
`Type PControl As Control Ptr`

**Comment by Kr00l** [comment]
I would suggest a shortcut version of a UDT declare.
Example
`Public Type HWND As Long`
So it can be Private or Public. The End Type is omitted and the datatype follows the type alias directly.

**Comment by wqweto** [comment]
Hiding pointers in typedef looks very powerful feature as well. Variables of `PControl` type above can be assigned LongPtr values say from VarPtr(MyUdt) but still dereferenced with members from `Control` e.g. pMyCtl.ControlMember1

This would make implementing WndProc and most other OS provided callbacks a breeze without CopyMemory and all the inefficient cruft VBx resorts to.

Such pointer types should be supported alias types in public typelibs though not sure.

**Comment by bclothier** [comment]
FWIW, I quite much like Kr00l's suggested syntax; feels like a natural extension of the existing `Type` statement. We still need to answer how the typecasting should be handled.

First, we need to recognize that in Windows it's very common to derive a type from another derived type. In Windows, the `HWND` is derived thus:

```
typedef void *PVOID;
typedef PVOID HANDLE;
typedef HANDLE HWND;
```

I am not sufficiently familiar with C/C++ to say whether that means the C/C++ compiler will allow downcasting of `HWND` into `HANDLE`, `HANDLE` into `PVOID`. If it does, then we need to allow for this with the proposed syntax:

```
Public Type PVOID As LongPtr
Public Type HANDLE As PVOID
Public Type HWND As HANDLE
Public Type HMENU As HANDLE
```

Given the syntax, this would be illegal:
```
Dim foo As HWND
Dim bar As HMENU

foo = bar 'Illegal, casting required
```

But this would be legal:
```
Dim foo As HWND
Dim bar As HANDLE

foo = bar 'Legal
bar = foo 'Not legal, casting required
```

We need to then consider what syntax for casting would be. In VB.NET they use `DirectCast`, `TryCast` and `CType`. All ugly, IMO. I think we should simply do something like this:
```
Dim foo As HANDLE 
Dim bar As HWND

bar = foo As HWND 'legal; casting is specified
```

In regards to @wqweto 's comment about callbacks --- I think that requires delegation support to properly handle the enforcement of procedure's signature when defining a callback procedure and may be a slightly different request. I love that idea, though as htat would make it very clear whether a procedure must be formatted a particular way.

**Comment by wqweto** [comment]
@bclothier VBx (COM actually) has already existing semantics for aliases (typedefs) and there so no (down) casting involved just regular conversion semantics for OLE Authomation types. 

There are no new types outside oleauth introduced per se unless hidden pointers above are implemented. Great opportunity IMO with not much effort.

**Comment by bclothier** [comment]
@wqweto , correct. However, I was looking more from the C/C++ compiler's viewpoint -- it's my understanding (but I may be mistaken) that they enforce casting of the aliased types, so it's more than just merely an alias. If it's merely an alias, then there's not much benefit and increased cognitive load. 

In Automation's case, [if I'm reading the specs correctly](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-oaut/7b5fa59b-d8f6-4a47-9695-630d3c10363e), aliases are not supposed to be used in an Automation-compatible interface. We would probably see them in an non-Automation type library, though. I think that's why VB* has been able to get away with not supporting the aliases.

**Comment by wqweto** [comment]
Aliases are “resolved” to their basic types before any marshaling or other oleauth “activity” so there is no problem to assign 123 to OLE_COLOR as it’s just an alias for Long.

**Comment by WaynePhillipsEA** [comment]
> I am not sufficiently familiar with C/C++ to say whether that means the C/C++ compiler will allow downcasting of `HWND` into `HANDLE`, `HANDLE` into `PVOID`. If it does, then we need to allow for this with the proposed syntax:

In C/C++, typedefs are treated exactly as if they are of the underlying type, and so in the example you gave, the values can be interchanged without any form of casting whatsoever.  Therefore, my stance is that typedef support should be implemented exactly as it exists already in external type libraries, and LongPtr casting should be considered as a separate feature request, where the syntax can be discussed separately.

**Comment by bclothier** [comment]
Thanks for the clarification. That said, I still would like to see compile-time checking of casting of typedefs (perhaps as a part of `Option Strict`) --- compile-time errors are much much cheaper to fix than runtime errors.

**Comment by WaynePhillipsEA** [comment]
> Thanks for the clarification. That said, I still would like to see compile-time checking of casting of typedefs (perhaps as a part of `Option Strict`) --- compile-time errors are much much cheaper to fix than runtime errors.

That sounds like a good compromise to me.

**Comment by EduardoVB** [comment]
> I would suggest a shortcut version of a UDT declare.
> Example
> `Public Type HWND As Long`

It sounds good, but I would suggest an another option to consider:

`Public Alias HWND As Long`

**Comment by WaynePhillipsEA** [comment]
> > I would suggest a shortcut version of a UDT declare.
> > Example
> > `Public Type HWND As Long`
> 
> It sounds good, but I would suggest another option to consider:
> 
> `Public Alias HWND As Long`

I really like both versions!  Let's put it to a vote...

**Comment by WaynePhillipsEA** [comment]
`Public Type HWND As Long`

If you prefer this syntax, please +1 this

**Comment by WaynePhillipsEA** [comment]
`Public Alias HWND As Long`

If you prefer this syntax, please +1 this

**Comment by DaveInCaz** [comment]
(Are there generics in TB?) How would this play with generics?

**Comment by WaynePhillipsEA** [comment]
> (Are there generics in TB?) How would this play with generics?

I can't foresee any problems relating to generics.  The compiler support is already included (in the form of typedef/aliases in type libraries), so it should just be a case of agreeing on the syntax.

**Comment by Kr00l** [comment]
> Whilst I like the Alias version, it would involve introducing a new keyword, and that keyword could potentially conflict with other syntax depending on the deceleration scope that we allow.
> 
> So we'll go with the familiar Type syntax, and now we just need to decide on the allowed declaration scope. I suggest we keep it simple and allow only component scope (i.e. module/class/interface). Later we might want to allow procedure scope as well, but that would require some resolver changes which would require a bit more work.

A typedef per procedure? So like a Dim?
Would this, in future, also allow a UDT per procedure scope?
E.g. `Dim Type HWND As Long`
Or
```
Dim Type PT
X As Long
Y As Long
End Type
```

**Comment by WaynePhillipsEA** [comment]
> > Whilst I like the Alias version, it would involve introducing a new keyword, and that keyword could potentially conflict with other syntax depending on the deceleration scope that we allow.
> > So we'll go with the familiar Type syntax, and now we just need to decide on the allowed declaration scope. I suggest we keep it simple and allow only component scope (i.e. module/class/interface). Later we might want to allow procedure scope as well, but that would require some resolver changes which would require a bit more work.
> 
> A typedef per procedure? So like a Dim?
> Would this, in future, also allow a UDT per procedure scope?
> E.g. `Dim Type HWND As Long`
> Or
> 
> ```
> Dim Type PT
> X As Long
> Y As Long
> End Type
> ```

Yes, that's the idea.  But I don't think we'd need the 'Dim ' prefix... since VBx doesn't allow use of the Type keyword inside the procedure scope, we can probably just use the normal Type syntax.   Thinking about this some more, we already have procedure-scoped type support, in the form of generics, so to implement this within procedure-scope should actually be much easier than first anticipated.

**Comment by Kr00l** [comment]
> > > Whilst I like the Alias version, it would involve introducing a new keyword, and that keyword could potentially conflict with other syntax depending on the deceleration scope that we allow.
> > > So we'll go with the familiar Type syntax, and now we just need to decide on the allowed declaration scope. I suggest we keep it simple and allow only component scope (i.e. module/class/interface). Later we might want to allow procedure scope as well, but that would require some resolver changes which would require a bit more work.
> > 
> > 
> > A typedef per procedure? So like a Dim?
> > Would this, in future, also allow a UDT per procedure scope?
> > E.g. `Dim Type HWND As Long`
> > Or
> > ```
> > Dim Type PT
> > X As Long
> > Y As Long
> > End Type
> > ```
> 
> Yes, that's the idea. But I don't think we'd need the 'Dim ' prefix... since VBx doesn't allow use of the Type keyword inside the procedure scope, we can probably just use the normal Type syntax. Thinking about this some more, we already have procedure-scoped type support, in the form of generics, so to implement this within procedure-scope should actually be much easier than first anticipated.

Skipping `Dim` is nicer.

I guess (?) it would be a niche, edge case use case.
Since UDT, typedef is used mostly with API declares it would need a wider scope than procedure anyway.
Or a UDT is also mostly used in a helper procedure receiving a ByRef UDT as Arg or return value.
However, if it's doable in a straight forward way, why not. :)

**Comment by EduardoVB** [comment]
Just to consider (probably not a good idea): how about adding procedure level API declaration capability?

**Comment by mwolfe02** [comment]
@bclothier wrote:

> Thanks for the clarification. That said, I still would like to see compile-time checking of casting of typedefs (perhaps as a part of `Option Strict`) --- compile-time errors are much much cheaper to fix than runtime errors.

Hear! Hear! [Some Bugs are Better Than Others](https://nolongerset.com/some-bugs-are-better-than-others/)

**Comment by Kr00l** [comment]
Is it possible (once feature is in place), when hovering over a "typedef", to show the "original" datatype.
Example:
```
Private Type HWND As LongPtr
Dim Handle As HWND
```
So, when hovering over the `As HWND` in the `Dim Handle` line to show LongPtr.

Ps: if this is already foreseen just ignore.
I got inspired by this at vbForums where a user said it is cumbersome sometimes to find the original datatype by all these aliases in c++.

**Comment by WaynePhillipsEA** [comment]
@Kr00l yes, it should do that automatically, but I will check once implemented.  Thanks

**Comment by sancarn** [comment]
Type aliasing is something I've wanted in a long time. This would make declaration of Win32 API libraries a hella lot easier.

**Comment by WaynePhillipsEA** [comment]
Yes, I think so too.  I'll bump this up the list to be done soon, as it will be an easy addition

**Comment by wqweto** [comment]
This 
```
    typedef struct {
        long   cbStruct;
    } MYSTRUCT;
    
    typedef [public] MYSTRUCT *LPMYSTRUCT;
```
. . . is some valid IDL (i.e. can be compiled to a TLB) which begs for this TB code
```
    Dim u As MYSTRUCT
    Dim p As LPMYSTRUCT
    
    u.cbStruct = 15
    p = VarPtr(u)
    Debug.Print p.cbStruct '--- should print 15
```
. . . to compile ok.

Currently VBx fails on LPMYSTRUCT with "Automation type not supported in Visual Basic".
